### PR TITLE
[proposal] Resolve Firebase auth in iOS

### LINKED ISSUE #353: DI framework for iOS targets
## Kind (Required)
- Proposal

 ## Overview (Required)
Why not use a DI framework for iOS targets?
Koin and Kodein are compatible with KMM.
As it is now, I think Hilt is fine for Android target. 
Because this app has to learn to experience Hilt as well.

 ## Links	
- https://insert-koin.io
- https://github.com/Kodein-Framework/Kodein-DI

### PR DESCRIPTION
**Pull request by watanavex**
## Issue
- close #353

## Overview (Required)
- This is a continuation of #396
- Firebase Auth abstracts and injects from Swift code
  - I was unable to link FirebaseAuth from the CocoaPods plugin in gradle.
  - GitLiveApp/firebase-kotlin-sdk uses Carthage
  - However, Firebase may discontinue support for Carthage.
  - So I thought of a way to abstract Firebase Auth and inject it from Swift code
- If you have another idea, it's welcome!

## Links
- https://github.com/GitLiveApp/firebase-kotlin-sdk/issues/133
- https://github.com/firebase/firebase-ios-sdk/blob/master/Carthage.md

## Screenshot
🙅‍♂️

### CODE REVIEW

**Comment by takahirom** [comment]
👀

File: ios-framework/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    val iosTarget: (String, KotlinNativeTarget.() -> Unit) -> KotlinNativeTarget =
+        if (System.getenv("SDK_NAME")?.startsWith("iphoneos") == true)
+            ::iosArm64
+        else
+            ::iosX64
+
+    iosTarget("ios") {
+        binaries {
+            framework {
+                baseName = "Shared"
```
Comment (takahirom) [inline ios-framework/build.gradle.kts:23]: Thank you for creating iOS combine framework! 👍 
How about "DroidKaigiMPP" for framework name 👀

**Comment by takahirom** [comment]
Will we implement Authenticator on the iOS side in the future? 👀

**Comment by watanavex** [comment]
> Will we implement Authenticator on the iOS side in the future? 👀

Exactly.
Sorry for not explaining enough.
We will call this method from Swift.
As a parameter, we pass the Authenticator implemented in Swift.

https://github.com/DroidKaigi/conference-app-2021/blob/484e986ba39a23650e1a1e0729509a6af67a0117/ios-framework/src/iosMain/kotlin/io/github/droidkaigi/feeder/IosModule.kt#L10-L21

File: ios-framework/build.gradle.kts (hunk 1 above)
Comment (watanavex) [reply ios-framework/build.gradle.kts:23]: Sorry, I changed my name.
I'll change it back!

https://github.com/DroidKaigi/conference-app-2021/blob/484e986ba39a23650e1a1e0729509a6af67a0117/ios/project.yml#L52-L53

**Comment by watanavex** [comment]
to-do
- Coroutines is locked to single-threaded version; change to multi-threaded version because Ktor doesn't work
- Implement Suspend Wrapper.
- Implement Authenticator in Swift

I'm expecting some problems with multithreading... :-x

File: ios-framework/build.gradle.kts (hunk 1 above)
Comment (watanavex) [reply ios-framework/build.gradle.kts]: Fixed 56e3f27 🙇‍♂️

**Review by takahirom — APPROVED** [review]
LGTM 👍 Thank you for advancing DroidKaigi!!

**Pull request merged**